src/structure.js

```diff
@@ -487,21 +492,31 @@ function hasIn(cursor, path) {
   return cursor.getIn(path, NOT_SET) !== NOT_SET;
 }
 
-function pathString(path) {
-  var topLevel = 'global';
-  if (!path || !path.length) return topLevel;
-  return [topLevel].concat(path).join('|');
+function listListenerMatching (listeners, path) {
+  var pathArray = [],
+      matches = [];
+
+  path = path || [];
+
+  path.forEach(function(pathPart) {
+    pathArray = pathArray.concat(pathPart);
+    matches = matches.concat(getListenerNs(listeners, pathArray));
+  });
+
+  return matches;
 }
 
-function listListenerMatching (listeners, basePath) {
-  var newListeners = [];
-  for (var key in listeners) {
-    if (!listeners.hasOwnProperty(key)) continue;
-    if (basePath.indexOf(key) !== 0) continue;
-    newListeners.push(listeners[key]);
+function getListenerNs (listeners, path) {
+  path = path || [];
+  path = path.concat("__listeners__");
```

Comment:
If you run the following code and look at the log you see the tree uses keys to store path structure, but also needs a place to put the callbacks so this is how I'm trying to do that and avoid any collisions.
```javascript
var struct = immstruct({foo: {bar: {baz: 1}}});
struct.reference(['foo', 'bar']).observe('swap', function() {console.log(arguments)});
struct.reference('foo').observe('swap', function() {console.log(arguments)});
console.log(struct._pathListeners);
```